core/src/test/java/org/apache/iceberg/deletes/TestRoaringPositionBitmap.java

```diff
@@ -0,0 +1,323 @@
+    bitmap1.add((long) 0 << 32 | 100L); // High = 0, Low = 100
+    bitmap1.add((long) 1 << 32 | 200L); // High = 1, Low = 200
+
+    RoaringPositionBitmap bitmap2 = new RoaringPositionBitmap();
+    bitmap2.add((long) 2 << 32 | 300L); // High = 2, Low = 300
+    bitmap2.add((long) 3 << 32 | 400L); // High = 3, Low = 400
+
+    bitmap1.or(bitmap2);
+
+    assertThat(bitmap1.contains((long) 0 << 32 | 100L)).isTrue();
+    assertThat(bitmap1.contains((long) 1 << 32 | 200L)).isTrue();
+    assertThat(bitmap1.contains((long) 2 << 32 | 300L)).isTrue();
+    assertThat(bitmap1.contains((long) 3 << 32 | 400L)).isTrue();
+
+    assertThat(bitmap1.cardinality()).isEqualTo(4);
+  }
+
+  @Test
+  public void testCardinality() {
+    RoaringPositionBitmap bitmap = new RoaringPositionBitmap();
+
+    assertThat(bitmap.cardinality()).isEqualTo(0);
+
+    bitmap.add(10L);
+    bitmap.add(20L);
+    bitmap.add(30L);
+
+    assertThat(bitmap.cardinality()).isEqualTo(3);
+
+    bitmap.add(10L); // already exists
+
+    assertThat(bitmap.cardinality()).isEqualTo(3);
+  }
+
+  @Test
+  public void testCardinalitySparseBitmaps() {
+    RoaringPositionBitmap bitmap = new RoaringPositionBitmap();
+
+    bitmap.add((long) 0 << 32 | 100L); // high = 0, low = 100
+    bitmap.add((long) 0 << 32 | 101L); // high = 0, low = 101
+    bitmap.add((long) 0 << 32 | 105L); // high = 0, low = 101
+    bitmap.add((long) 1 << 32 | 200L); // high = 1, low = 200
+    bitmap.add((long) 100 << 32 | 300L); // high = 1000, low = 300
+
+    assertThat(bitmap.cardinality()).isEqualTo(5);
+  }
+
+  @Test
+  public void testSerializeDeserializeAllContainerBitmap() {
+    RoaringPositionBitmap bitmap = new RoaringPositionBitmap();
+
+    // bitmap 0, container 0 (array)
+    bitmap.add(position(0 /* bitmap */, 0 /* container */, 5L));
+    bitmap.add(position(0 /* bitmap */, 0 /* container */, 7L));
+
+    // bitmap 0, container 1 (array that can be compressed)
+    bitmap.addRange(
+        position(0 /* bitmap */, 1 /* container */, 1L),
+        position(0 /* bitmap */, 1 /* container */, 1000L));
+
+    // bitmap 1, container 2 (bitset)
+    bitmap.addRange(
+        position(0 /* bitmap */, 2 /* container */, 1L),
+        position(0 /* bitmap */, 2 /* container */, CONTAINER_OFFSET - 1L));
+
+    // bitmap 1, container 0 (array)
+    bitmap.add(position(1 /* bitmap */, 0 /* container */, 10L));
+    bitmap.add(position(1 /* bitmap */, 0 /* container */, 20L));
+
+    // bitmap 1, container 1 (array that can be compressed)
+    bitmap.addRange(
+        position(1 /* bitmap */, 1 /* container */, 10L),
+        position(1 /* bitmap */, 1 /* container */, 500L));
+
+    // bitmap 1, container 2 (bitset)
+    bitmap.addRange(
+        position(1 /* bitmap */, 2 /* container */, 1L),
+        position(1 /* bitmap */, 2 /* container */, CONTAINER_OFFSET - 1));
+
+    assertThat(bitmap.runOptimize()).as("Bitmap must be RLE encoded").isTrue();
```

Comment:
Should optimize be run as part of serialization?

Reply:
I [call that](https://github.com/apache/iceberg/pull/11302/files#diff-fb7fc774bdf5fb14c7e9b4aab2ed645f5fcdafc7d98cce8619134cce53b4d931R126) in `PositionDeleteIndex`, not sure about the bitmap itself.

Reply:
Actually, we can't do that. The caller may have called `serializedSizeInBytes()` to estimate the size prior to the serialization.